### PR TITLE
fix(explore): adhoc metrics popover resets label after hovering outside

### LINKED ISSUE #16067: [explore] "Unexpected Error"/chart not render on SUM metric
Get an unexpected error when trying to generate a superset chart (as seen in the screenshot below) even though the query runs and returns results.

#### Screenshots
image-2021-08-02-15-19-23-281 (1)

You will see that `data` section has correct results but chart can't be rendered.

It seems like the SUM metrics are incorrectly being named "sum" in the generated query, which causes some weird sql problems.
image-2021-08-02-15-21-43-760

Switching to "custom sql" is a workaround but not ideal.

### Environment

latest production

### Checklist

Make sure to follow these steps before submitting your issue - thank you!

- [x] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [x] I have reproduced the issue with at least the latest released version of superset.
- [x] I have checked the issue tracker for the same issue and I haven't found one similar.

### Additional context

CC @serenajiang @villebro

### PR DESCRIPTION
**Pull request by kgabryje**
### SUMMARY
When user opened "Add new metric" popover and selected a column and/or aggregate, metric title is automatically set to "column(aggregate)". However, when user hovers with their mouse outside of the popover before saving, the title gets reset. Then, after saving, the label appears to be correct, but an empty string gets passed to executed query as a label, causing seemingly unrelated errors.
This PR converts `MetricsControl` component from class to functional and uses memoization to prevent resetting the title.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Before: see linked issue and video in the comments
After:
https://user-images.githubusercontent.com/15073128/129045681-e33f2ab4-4aac-4d95-b928-6f5c3705e13f.mov

### TESTING INSTRUCTIONS
1. Open a chart
2. Open new metric popover
3. Add a column and aggregate
4. Hover with your mouse outside of the popover - nothing should happen
5. Save the metric and run query
6. Chart should render as expected

### ADDITIONAL INFORMATION

- [x] Has associated issue: fixes #16067 
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

CC @graceguo-supercat @junlincc

### CODE REVIEW

File: superset-frontend/src/explore/components/controls/MetricControl/MetricsControl.jsx [hunk 1]
```diff
@@ -123,295 +116,233 @@ function coerceAdhocMetrics(value) {
...
-              ? value.optionName === oldMetric.optionName
-              : false
-          ) {
-            return changedMetric;
+      let transformedOpts;
+      if (Array.isArray(opts)) {
+        transformedOpts = opts;
+      } else {
+        transformedOpts = opts ? [opts] : [];
+      }
```
Comment (villebro) [inline superset-frontend/src/explore/components/controls/MetricControl/MetricsControl.jsx:147]: nit: would `const transformedOpts = ensureIsArray(opts)` work here?

**Review by villebro — APPROVED** [review]
Code LGTM with minor nit. I tried to make this break in every possible way I could but wasn't able to come up with any bugs/regressions. Extra kudos for net negative LOC PR!

File: superset-frontend/src/explore/components/controls/MetricControl/MetricsControl.jsx (hunk 1 above)
Comment (kgabryje) [reply superset-frontend/src/explore/components/controls/MetricControl/MetricsControl.jsx]: Good point, done!

**Review by mistercrunch — APPROVED** [review]
I pulled the branch and tried to reproduce the bug that I had become very acquainted with last night and could not reproduce. Tried with flags `ENABLE_EXPLORE_DRAG_AND_DROP` and `UX_BETA` on and off and couldn't repro the bug or didn't find any other issues.

**Comment by rosemarie-chiu** [comment]
🏷 2021.31

**Pull request merged**